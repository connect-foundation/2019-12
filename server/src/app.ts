import * as express from 'express';

const app = express();

export default app;
