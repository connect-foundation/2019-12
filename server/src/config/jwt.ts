export const JWT_SECRET = '1234';
export const JWT_EXPIRE = '30m';
export const JWT_SUBJECT = 'USER';
