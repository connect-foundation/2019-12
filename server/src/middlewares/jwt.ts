import * as JWT from 'jsonwebtoken';

export function generateJWT(
  id: number,
  googleId: number,
  email?: string,
): Promise<JWT.Secret> {
  return new Promise((resolve, reject) => {
    JWT.sign(
      {
        id,
        googleId,
        email,
      },
      'secret',
      {
        // Expire Date를 외부로 빼야함.
        expiresIn: 1000 * 60 * 60 * 24,
      },
      (err, token) => {
        if (err) reject(err);
        resolve(token);
      },
    );
  });
}
