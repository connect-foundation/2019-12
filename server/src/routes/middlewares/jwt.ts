import * as express from 'express';
//import * as jwt from 'jsonwebtoken';

//import { JWT_SECRET, JWT_EXPIRE, JWT_SUBJECT } from '../../config/jwt';


function setJWT(payload) {
  return new Promise((resolve, reject) => {
    jwt.sign(
      payload,
      JWT_SECRET,
      {
        expiresIn: JWT_EXPIRE,
        subject: JWT_SUBJECT,
      },
      (err, token) => {
        if (err) reject(err);
        resolve(token);
      },
    );
  });
}
export async function makeToken(
  req: express.Request,
  res: express.Response,
  next: express.NextFunction,
) {
  try {
    // const payload = req.user;
    //req.user 객체에 id와 token 정보가 넘어올 예정임. 추후 obj 모양을 정의할 것임.
    // let token = await setJWT(payload);
    // res.cookie('auth', token, { maxAge: 1000 * 60 * 60 });
    next();
  } catch (err) {
    res.status(500).send(err);
  }
}
