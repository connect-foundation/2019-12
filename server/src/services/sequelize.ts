import { Sequelize } from 'sequelize-typescript';
import {
  Event,
  Order,
  OrderTicket,
  TicketSubscription,
  TicketType,
  User,
} from '../models';

const { DB_HOST, DB_USER, DB_PW, DB_NAME } = process.env;

export const sequelize = new Sequelize({
  host: DB_HOST,
  username: DB_USER,
  password: DB_PW || '',
  database: DB_NAME,
  dialect: 'mariadb',
  models: [Event, Order, OrderTicket, TicketSubscription, TicketType, User],
});
