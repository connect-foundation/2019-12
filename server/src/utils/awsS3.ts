import { S3, config } from 'aws-sdk';
import { ReadStream } from 'fs';

const {
  AWS_ACCESS_KEY: accessKeyId,
  AWS_SECRET_KEY: secretAccessKey,
  AWS_S3_ENDPOINT: endpoint,
  AWS_REGION: region,
  AWS_BUCKET_NAME: Bucket,
} = process.env;

config.update({ accessKeyId, secretAccessKey });
const s3 = new S3({ region, endpoint });

interface PutObjectResult {
  URL: string;
  ETag?: string;
}

export async function putObject(
  Key: string,
  Body: ReadStream | Buffer,
): Promise<PutObjectResult> {
  if (!Bucket) throw Error('Bucket Not Found');
  const { ETag } = await s3
    .putObject({
      Bucket,
      Key,
      Body,
      ACL: 'public-read',
    })
    .promise();
  const URL = `${endpoint}/${Bucket}/${Key}`;

  return { URL, ETag };
}
