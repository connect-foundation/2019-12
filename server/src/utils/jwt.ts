import * as JWT from 'jsonwebtoken';
import { tokentype } from '../types';
const { JWT_SECURE } = process.env;

export function generateJWT(
  exist: boolean,
  id: number,
  googleId: number,
  email: string,
): Promise<JWT.Secret> {
  const payload: tokentype = {
    exist,
    id,
    googleId,
    email,
  };
  return new Promise((resolve, reject) => {
    JWT.sign(
      payload,
      JWT_SECURE || '',
      {
        expiresIn: 1000 * 60 * 60 * 24,
      },
      (err, token) => {
        if (err) reject(err);
        resolve(token);
      },
    );
  });
}
