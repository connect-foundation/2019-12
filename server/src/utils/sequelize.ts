import { Sequelize } from 'sequelize-typescript';
import { Event, UserTicket, TicketType, User } from '../models';
import { readJSONData } from './readJSON';
import { resolve } from 'path';

const { DB_HOST, DB_PORT, DB_USER, DB_PW, DB_NAME } = process.env;
const SEED_DIR = resolve(__dirname, '../../db_seeds');

export const sequelize = new Sequelize({
  host: DB_HOST,
  port: DB_PORT ? +DB_PORT : 3306,
  username: DB_USER,
  password: DB_PW,
  database: DB_NAME,
  dialect: 'mariadb',
  models: [Event, UserTicket, TicketType, User],
});

export async function migrate() {
  console.info('DB migration start...');
  await sequelize.sync();
  console.info('DB migration end...');
}

export async function seed() {
  console.info('DB seed start...');
  // UserData 추가
  const usersPath = resolve(SEED_DIR, 'users.json');
  const usersData = await readJSONData<User>(usersPath);
  await User.bulkCreate(usersData);

  // EventData 추가
  const eventPath = resolve(SEED_DIR, 'events.json');
  const eventData = await readJSONData<Event>(eventPath);
  await Event.bulkCreate(eventData);

  // TicketData 추가
  const ticketPath = resolve(SEED_DIR, 'ticketTypes.json');
  const ticketData = await readJSONData<TicketType>(ticketPath);
  await TicketType.bulkCreate(ticketData);

  // OrderTikcetData 추가
  const userTicketPath = resolve(SEED_DIR, 'userTickets.json');
  const userTicketData = await readJSONData<UserTicket>(userTicketPath);
  await UserTicket.bulkCreate(userTicketData);

  console.info('DB seed end...');
}
