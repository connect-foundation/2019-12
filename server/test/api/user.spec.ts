import * as request from 'supertest';
import app from '../../src/app';
import { Secret } from 'jsonwebtoken';
import { sequelize } from '../../src/utils/sequelize';
import { generateJWT } from '../../src/utils/jwt';
import { OK, UNAUTHORIZED, NO_CONTENT } from 'http-status';

beforeAll(async () => {
  sequelize.options.logging = false;
  await sequelize.sync();
});

afterAll(() => {
  sequelize.close();
});

describe('Router GET /api/users/tickets', () => {
  function setHeader(token: Secret) {
    return {
      Cookie: `UID=${token}`,
      Accept: 'application/json',
    };
  }

  it('로그인 안했을 경우', async () => {
    const token = await generateJWT(false, 1, 1, '1234@gmail.com');
    await request(app)
      .get('/api/users/tickets')
      .set(setHeader(token))
      .expect(UNAUTHORIZED);
  });
  it('유저 데이터를 불러왔고, 데이터가 있을 경우', async () => {
    const token = await generateJWT(true, 2, 1, '1234@gmail.com');
    await request(app)
      .get('/api/users/tickets')
      .set(setHeader(token))
      .expect(OK)
      .expect(res => {
        const data = res.body[0];
        const { userTickets, ...body } = data;
        const { createdAt, ...userTicket } = userTickets[0];
        expect(body).toMatchSnapshot();
        expect(userTicket).toMatchSnapshot();
      });
  });
  it('유저 데이터를 불러왔고, 데이터가 없을 경우 204', async () => {
    const token = await generateJWT(true, 1, 1, '1234@gmail.com');
    await request(app)
      .get('/api/users/tickets')
      .set(setHeader(token))
      .expect(NO_CONTENT)
      .expect(res => {
        expect(res.body).toStrictEqual({});
      });
  });
});

describe('Router GET /api/users/events', () => {
  function setHeader(token: Secret) {
    return {
      Cookie: `UID=${token}`,
      Accept: 'application/json',
    };
  }

  it('로그인 안했을 경우', async () => {
    const token = await generateJWT(false, 1, 1, '1234@gmail.com');
    await request(app)
      .get('/api/users/events')
      .set(setHeader(token))
      .expect(UNAUTHORIZED);
  });
  it('유저의 이벤트를 불러왔고, 데이터가 있을 경우 200', async () => {
    const token = await generateJWT(true, 2, 1, '1234@gmail.com');
    await request(app)
      .get('/api/users/events')
      .set(setHeader(token))
      .expect(OK);
  });
  it('유저의 이벤트를 불러왔고, 데이터가 없을 경우 204', async () => {
    const token = await generateJWT(true, 100000, 1, '1234@gmail.com');
    await request(app)
      .get('/api/users/events')
      .set(setHeader(token))
      .expect(NO_CONTENT)
      .expect(res => {
        expect(res.body).toStrictEqual({});
      });
  });
});
