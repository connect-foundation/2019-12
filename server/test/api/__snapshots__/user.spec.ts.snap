// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Router GET /api/users/tickets 유저 데이터를 불러왔고, 데이터가 있을 경우 1`] = `
Object {
  "address": "서울특별시 강남구 역삼동 736-40 아름다운 빌딩 6층 <복합문화공간 위플레이스>",
  "endAt": "2018-03-09T12:30:00.000Z",
  "id": 2,
  "mainImg": "https://bookus.kr.object.ncloudstorage.com/10",
  "place": "역삼역 위플레이스 다목적홀",
  "placeDesc": "",
  "startAt": "2018-03-09T10:30:00.000Z",
  "ticket": Object {
    "desc": "",
    "eventId": 2,
    "id": 2,
    "isPublicLeftCnt": false,
    "leftCnt": 10,
    "maxCntPerPerson": 1,
    "name": "선착순 (참가확정)",
    "price": 0,
    "quantity": 30,
    "refundEndAt": "2200-03-06T10:30:00.000Z",
    "salesEndAt": "2200-03-06T10:30:00.000Z",
    "salesStartAt": "2018-03-02T06:00:00.000Z",
  },
  "title": "LTNS, 9XD! ; 10회 모임 @ZOYI Corporation",
  "userId": 2,
}
`;

exports[`Router GET /api/users/tickets 유저 데이터를 불러왔고, 데이터가 있을 경우 2`] = `
Object {
  "id": 1,
  "isAttendance": true,
  "ticketTypeId": 2,
  "userId": 2,
}
`;
